### PR TITLE
Feature logger

### LINKED ISSUE #12: Implement a (proper) logger

### PR DESCRIPTION
**Pull request by smellilac**

### CODE REVIEW

**Comment by smellilac** [comment]
Git asking to add to name of "LogMessage" method, because because it sees Task in the method return parameter, but I use 
Task.CompletedTask; so it doesn't make sense

**Comment by HEJOK254** [comment]
> Git asking to add to name of "LogMessage" method, because because it sees Task in the method return parameter, but I use 
> Task.CompletedTask; so it doesn't make sense

There seems to be a lot of debate in the c# community regarding the naming of async functions. I chose to go with Microsoft's guidelines, and add the Async suffix to the method names, as DeepSource tells me to do.

The internet has different opinions about adding the Async suffix to Tasks without the async modifier aswell. I'd rename it to LogMessageAsync, but I'm not sure how others like it. I'm not really sure what to do here to be honest

**Comment by smellilac** [comment]
> > Git asking to add to name of "LogMessage" method, because because it sees Task in the method return parameter, but I use
> > Task.CompletedTask; so it doesn't make sense
> 
> There seems to be a lot of debate in the c# community regarding the naming of async functions. I chose to go with Microsoft's guidelines, and add the Async suffix to the method names, as DeepSource tells me to do.
> 
> The internet has different opinions about adding the Async suffix to Tasks without the async modifier aswell. I'd rename it to LogMessageAsync, but I'm not sure how others like it. I'm not really sure what to do here to be honest

In this case it is not necessary, since the method does not perform asynchronous operations.
I think it would be more misleading than reflective of the asynchronicity.

**Comment by HEJOK254** [comment]
> In this case it is not necessary, since the method does not perform asynchronous operations.
> I think it would be more misleading than reflective of the asynchronicity.

In this case, the issue can be ignored by adding a comment in the same line, or the line above: `skipcq: CS-R1073`

**Comment by smellilac** [comment]
> > In this case it is not necessary, since the method does not perform asynchronous operations.
> > I think it would be more misleading than reflective of the asynchronicity.
> 
> In this case, the issue can be ignored by adding a comment in the same line, or the line above: `skipcq: CS-R1073`

It was good idea :)

**Comment by HEJOK254** [comment]
Sorry for the delay, I'm going to try to review the pr today and suggest some changes if needed

File: Program.cs
Comment (HEJOK254) [inline Program.cs]: ShowStartMessage() should be enough for the user to know that the program started, so the information log could be changed to a debug log.

The current approach is fine too, just looks a bit weird with a log message before the startup message

File: Commands/InteractionServiceHandler.cs
Comment (HEJOK254) [inline Commands/InteractionServiceHandler.cs]: Provide additional information about the error
```suggestion
		catch (Exception e)
```

File: Commands/InteractionServiceHandler.cs
Comment (HEJOK254) [inline Commands/InteractionServiceHandler.cs]: Provide additional information about the error

```suggestion
			Log.Fatal($"Error initializing InteractionService: {e.Message}");
```

File: Commands/InteractionServiceHandler.cs
Comment (HEJOK254) [inline Commands/InteractionServiceHandler.cs]: I think providing details about where the error comes from is a bit unnecessary, I'm not sure why I did that really lol
What do you think?

```suggestion
			Log.Error("Failed to register modules: InteractionService not initialized.");
```

File: Commands/InteractionServiceHandler.cs
Comment (HEJOK254) [inline Commands/InteractionServiceHandler.cs]: Same situation as above
It could also be simplified to just "Registered modules", but I'm not sure if the "Modules registered successfully" isn't better, as it's clearly indicating a success

```suggestion
			Log.Information("Registered modules");
```
or
```suggestion
			Log.Information("Modules registered successfully");
```

File: Commands/InteractionServiceHandler.cs
Comment (HEJOK254) [inline Commands/InteractionServiceHandler.cs]: Not sure if the user needs the whole error.
Could be dependent on the debug bool from the config for example

```suggestion
			Log.Fatal($"Error registering modules: {e.Message}");
```
Or dependent on the debug bool from the config, though the current implementation would require quite a bit of code
```suggestion
						Log.Fatal($"Error registering modules: {(Program.config != null && Program.config.debug ? e : e.Message)}");
```
Or the additional information could be provided with a separate debug log, requiring less code, but the message would be split into 2.
All of this could be implemented into the other error logs

File: Commands/InteractionServiceHandler.cs
Comment (HEJOK254) [inline Commands/InteractionServiceHandler.cs]: Same as in `RegisterModulesAsync()`

```suggestion
			Log.Error("Error handling interaction: InteractionService not initialized.");
```

File: Commands/InteractionServiceHandler.cs
Comment (HEJOK254) [inline Commands/InteractionServiceHandler.cs]: Just a small change for better log message consistency

```suggestion
			Log.Error($"Error handling interaction: {e.Message}");
```

File: Commands/Modules/VideoUtils.cs
Comment (HEJOK254) [inline Commands/Modules/VideoUtils.cs]: Same thing as in `InteractionServiceHandler.cs` basically. Especially since this is normally a one time message, and it's only present in one class.

```suggestion
			Log.Information("TMP directory not found. Created it automatically");
```

File: ConfigManager.cs
Comment (HEJOK254) [inline ConfigManager.cs]: The message doesn't actually return a path, but just says "./config.json", which doesn't help fix the problem
I think I actually forgot to test it in my original code, so I guess it can be fixed now :P

```suggestion
            Log.Fatal($"Config file not found at: {Path.GetFullPath(path)}");
```

File: ConfigManager.cs
Comment (HEJOK254) [inline ConfigManager.cs]: Provide more information

```suggestion
		catch (Exception e)
```

File: ConfigManager.cs
Comment (HEJOK254) [inline ConfigManager.cs]: Use with the change above
Provide more information about the issue, so the user can have a general idea of what the problem is

```suggestion
            Log.Fatal($"Failed to parse config file: {e.Message}");
```

File: LoggerConfig/LoggerImplementation/AutoLog.cs
Comment (HEJOK254) [inline LoggerConfig/LoggerImplementation/AutoLog.cs]: Explain why the error is skipped

```suggestion
    // LogMessage does not run asynchronously, so we can ignore the DeepSource error
    // skipcq: CS-R1073
```

File: LoggerConfig/LoggerImplementation/AutoLog.cs
Comment (HEJOK254) [inline LoggerConfig/LoggerImplementation/AutoLog.cs]: I think `AutoLog` sounds a bit weird, but I honestly can't think of anything else except for like `DiscordLogger`, which would kind of sound like it's logging actual messages from discord, which would sound a bit like the bot is "spying"

The namespace makes `AutoLog` sound better though, so I think it could be left like this

File: LoggerConfig/LoggerImplementation/AutoLog.cs
Comment (HEJOK254) [inline LoggerConfig/LoggerImplementation/AutoLog.cs]: `LogMessage()` sounds a bit generic but once again, I can't think of any other good name, so I think it's fine

File: LoggerConfig/LoggerImplementation/AutoLog.cs
Comment (HEJOK254) [inline LoggerConfig/LoggerImplementation/AutoLog.cs]: This class doesn't really have anything to do with configuration, so I think changing the namespace both here and in `SerilogConfiguration.cs` to something like `QuickEdit.Logger` or `QuickEdit.Logging` *could* be better? I'm not sure though

File: LoggerConfig/SerilogConfiguration.cs
Comment (HEJOK254) [inline LoggerConfig/SerilogConfiguration.cs]: The logger should probably log debug messages when debug is on in the configuration

I'm not very familiar with Serilog just yet, so my implementation of this might not be very good, but this is what I came up with:
```suggestion
	var loggerConfig = new LoggerConfiguration()
		.Enrich.WithMachineName()
		.WriteTo.Console()
		.WriteTo.File(logPath, rollingInterval: RollingInterval.Day);

	if (Program.config != null && Program.config.debug)
		loggerConfig.MinimumLevel.Debug();
	else
		loggerConfig.MinimumLevel.Information();
		
	Log.Logger = loggerConfig.CreateLogger();
```

**Review by HEJOK254 — CHANGES_REQUESTED** [review]
I've reviewed the changes, and everything seems to be working great, however I made come suggestions here and there.

Keep in mind that these are just suggestions :)

File: Commands/InteractionServiceHandler.cs
Comment (smellilac) [reply Commands/InteractionServiceHandler.cs]: Yes, I also think it is unnecessary.

File: Commands/InteractionServiceHandler.cs
Comment (smellilac) [reply Commands/InteractionServiceHandler.cs]: The second option is of course voluminous in terms of code, but extremely correct in terms of its functionality, so it is necessary to choose it.

**Comment by smellilac** [comment]
You don't merge PR.
Want to add/correct something?

**Comment by HEJOK254** [comment]
> You don't merge PR.
> Want to add/correct something?

Sorry, I've been out of town for a few days, so I wasn't able to review it

I should be back in a few days, and I'll try to review it when I can

Sorry for the delay

File: QuickEdit.Logger/LoggerImplementation/AutoLog.cs
Comment (HEJOK254) [inline QuickEdit.Logger/LoggerImplementation/AutoLog.cs:1]: I meant the namespace in the file, not the folder name, sorry :/

So the folder would be called Logger (and the file path would be `/Logger/AutoLog.cs`), while the namespace on line 10 could be for example `QuickEdit.Logger`

File: QuickEdit.Logger/SerilogConfiguration.cs
Comment (HEJOK254) [inline QuickEdit.Logger/SerilogConfiguration.cs:1]: I meant the namespace in the file, not the folder name, sorry :/

So the folder would be called Logger (and the file path would be `/Logger/SerilogConfiguration.cs`), while the namespace on line 9 could be for example `QuickEdit.Logger`

**Review by HEJOK254 — CHANGES_REQUESTED** [review]
I've reviewed the changes, sorry for the delay

**Comment by smellilac** [comment]
> delay

No problem. Glad to have you back!

File: QuickEdit.Logger/LoggerImplementation/AutoLog.cs
Comment (smellilac) [reply QuickEdit.Logger/LoggerImplementation/AutoLog.cs:1]: > I meant the namespace in the file, not the folder name, sorry :/
> 
> So the folder would be called Logger (and the file path would be `/Logger/AutoLog.cs`), while the namespace on line 10 could be for example `QuickEdit.Logger`

It's okay, misunderstandings happen to everyone :relaxed:

I'll fix it now

**Comment by smellilac** [comment]
I renamed the main folder to "Logger" and deleted the "LoggerImplementation" folder.

I hope we understand each other correctly now, if not, please let me know.

File: Logger/SerilogConfiguration.cs
Comment (HEJOK254) [inline Logger/SerilogConfiguration.cs]: Use the same namespace as `AutoLog.cs`, as there isn't really a reason to separate the two
```suggestion
namespace QuickEdit.Logger;
```

File: Program.cs
Comment (HEJOK254) [inline Program.cs]: If `AutoLog` is moved to the `QuickEdit.Logger` namespace, the `QuickEdit.LoggerConfig` using can be removed
```suggestion
```

**Review by HEJOK254 — CHANGES_REQUESTED** [review]
It's all good now, but I think `AutoLog` and `SerilogConfiguration` should be in the same namespace, since they're only 2 classes, and they're related

**Review by HEJOK254 — DISMISSED** [review]
Looks good!

I'd clean up the commit history a bit by doing an interactive rebase if you're ok with that. It would basically rewrite the commit history, in such a way that, for example, some of the small commits fixing an issue with the previous commit, which are quite common when working on a feature, could be merged into one, cleaning up the history a bit before getting merged.

This would also require a force push, to rewrite the commits. I always make sure to keep a backup of the branch, though, in case something goes wrong.

After that, I would slightly clean up all the files, like for example fixing some broken indentations in some places probably caused by me forgetting to add them in the review suggestions :P

I'm not an expert at git, and this is pretty much my first time collaborating with other people on GitHub (thanks :)), which is why I'm always asking about everything. I can just leave the commit history as is and merge the PR, or squash merge it if needed, but I just thought it would be better to tidy it up a bit.

**Comment by smellilac** [comment]
> Looks good!
> 
> I'd clean up the commit history a bit by doing an interactive rebase if you're ok with that. It would basically rewrite the commit history, in such a way that, for example, some of the small commits fixing an issue with the previous commit, which are quite common when working on a feature, could be merged into one, cleaning up the history a bit before getting merged.
> 
> This would also require a force push, to rewrite the commits. I always make sure to keep a backup of the branch, though, in case something goes wrong.
> 
> After that, I would slightly clean up all the files, like for example fixing some broken indentations in some places probably caused by me forgetting to add them in the review suggestions :P
> 
> I'm not an expert at git, and this is pretty much my first time collaborating with other people on GitHub (thanks :)), which is why I'm always asking about everything. I can just leave the commit history as is and merge the PR, or squash merge it if needed, but I just thought it would be better to tidy it up a bit.

Yes, indeed, the PR looks dirty, so I'm all for you cleaning it up.

I'm also far from being a professional in git and this is also my first work in collaboration (thank you very much too!), but as far as I understand, I can't help with cleaning the commits, since I don't have the necessary rights in your repository. However, I can clean the code and upload the commit.

After which you will start cleaning, okay?

**Comment by HEJOK254** [comment]
> However, I can clean the code and upload the commit.

Yeah, you can do that if you want.
I was basically thinking of fixing the indentations and formatting, plus maybe removing the unused `using` directives. All of that can be achieved by running a clean-up in Visual Studio, as far as I'm aware, and additionally double-checking if everything is correct and works properly.

I can do it if you don't want to, and as for the history clean-up, I wanted to do it in this pull request, so on your fork, in your branch, as I think I should have the necessary permissions for it, but I'm not sure yet.

After that, I'll merge the pr

**Comment by smellilac** [comment]
> After that, I'll merge the pr

Ok, I didn't know it was so easy to clean up code from extra spaces and unused directives.

In that case, I'm waiting for your code cleaning, as well as cleaning up commits.

**Comment by HEJOK254** [comment]
I pushed the changes to [HEJOK254/Discord-QuickEdit/feature/logger-preview](https://github.com/HEJOK254/Discord-QuickEdit/tree/feature/logger-preview), instead of force pushing it to your branch, so we can discuss the changes and change things if needed, without having to force push on your branch all the time.

Please check if everything looks ok, or if I should rename/delete/squash etc. some commits. I tried to keep the commits small, which should help in troubleshooting in case one of them introduced some issue, while also grouping together changes regarding the same "feature", like for example improving the log messages.

If you have any suggestions, let me know, and I'll be happy to change it

**Comment by smellilac** [comment]
> If you have any suggestions, let me know, and I'll be happy to change it
I looked at the commits in this branch. They look clean and structured. I have no ideas/wishes, in my opinion you did everything well.

**Comment by smellilac** [comment]
> > If you have any suggestions, let me know, and I'll be happy to change it

 I looked at the commits in this branch. They look clean and structured. I have no ideas/wishes, in my opinion you did everything well.

File: Logger/SerilogConfiguration.cs
Comment (HEJOK254) [inline Logger/SerilogConfiguration.cs]: The log is saved in a location only appropriate in Visual Studio. On a server, it would create the logs in a hard to find path, and might cause confusion.

I'd recommend changing it to just `logs`, as it should be clear to the user where the logs are located.
That would also eliminate the need for the `intermediateOutputPath` variable, as just "logs" could be used, though a better approach, I think, is still using a variable, but renaming it.

Another thing I've noticed, is the log names contain the date twice, and are not really specifying what program they're from. From my testing, it seems that Serilog (or something else idk) automatically adds the date to the filename, removing the need for manually adding the date and time.

```suggestion
		var logDirectory = "logs";

		Directory.CreateDirectory(logDirectory);

		var logPath = Path.Combine(logDirectory, $"quickedit-.log");
```

File: Logger/SerilogConfiguration.cs
Comment (HEJOK254) [inline Logger/SerilogConfiguration.cs]: I don't think this enrichment is used at all here. To use it, I think a custom `outputTemplate` should be used. However, I don't think this feature is necessary in this case, so it could just be removed along with the `Serilog.Enrichers.Environment` NuGet package.

```suggestion
```

If you think it should stay, let me know

**Review by HEJOK254 — CHANGES_REQUESTED** [review]
While reviewing the code before merging, a few things have caught my eye, that I think should be fixed

Sorry I'm noticing about it just now

**Comment by HEJOK254** [comment]
I forgot to remove that `$` sign in `SerilogConfiguration.cs` in the review suggestion, which was causing the DeepSource issue, and I also forgot to add a suggestion to remove the Serilog Enricher, which is now unused, so I went ahead and committed the fixes myself, hope that's not an issue

Anyway, I think the PR is ready to be merged. If you have any questions or any more changes you would like to make, let me know, if not, I'll merge it soon :)

**Comment by smellilac** [comment]
> Anyway, I think the PR is ready to be merged. If you have any questions or any more changes you would like to make, let me know, if not, I'll merge it soon :)

No questions or suggestions.

Thank you for working together!

**Comment by HEJOK254** [comment]
> No questions or suggestions.

Ok, I'll merge it now

Thanks for helping me with the project :)

**Review by HEJOK254 — APPROVED** [review]

**Pull request merged**
